### PR TITLE
Feature request: ARP flush cache table #1833

### PR DESCRIPTION
**Pull request by peppelinux**
Very simple approach, it flushes every arp table.
If preferred we could introduce a select menu for selecting the interface of which the table is referred.

### CODE REVIEW

File: src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php [hunk 1]
```diff
@@ -74,7 +74,18 @@ public function getArpAction()
...
         return $arptable;
     }
-
+    
+    /**
+     * retrieve system arp table contents
+     * @return array
+     */
+    public function flushArpAction()
+    {
```
Comment (AdSchellevis) [inline src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php]: @peppelinux please validate for a post type request, and use ajaxCall in the volt template. this breaks csrf.

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 2]
```diff
@@ -61,7 +61,17 @@ POSSIBILITY OF SUCH DAMAGE.
...
+        
+        
+        function flushARP() {
+            ajaxGet(url = "/api/diagnostics/interface/flushArp",
+                sendData = {}, callback = function (data, status) {
+                    $("#refresh").click();
+                });
+        }
 
+        $("#flush").click(flushARP);
```
Comment (AdSchellevis) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: @peppelinux probably better to wrap all in the .click() action and add a dialog to warn the user, an arp flush can have side affects

File: src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php (hunk 1 above)
Comment (peppelinux) [reply src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php]: You're right!
I've just cloned the programming style of InterfaceController.php.
By the way, Are you asking me to put something like:
````
if ($this->request->isPost()) { ...
````
if it is it consider it done but is there any suggestion on csf token variable name to use in the template?

Comment (AdSchellevis) [reply src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php]: csrf headers are included and validated automatically,  the isPost() is indeed what I mean.

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt (hunk 2 above)
Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: I'm looking for a bootstrap class template or whatever builtin in OPN UI

Comment (AdSchellevis) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: you can use bootstrapdialog, for example https://github.com/opnsense/core/blob/c14000892b603a70c931aca44db01021e49d12d8/src/www/interfaces_groups.php#L71

Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: Ops, I was late to read it. I'll use this next time, very very thanks

File: src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php (hunk 1 above)
Comment (peppelinux) [reply src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php]: Fixed :+1:

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 3]
```diff
@@ -78,6 +78,43 @@ POSSIBILITY OF SUCH DAMAGE.
...
+
+    <!-- Modal content-->
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Flush ARP system table</h4>
+      </div>
+      <div class="modal-body">
+        <p>
+        If you've changed the IP of a host - or setup a new host with the IP of an old one - you've probably get that the host has no network connectivity for a period of time. The router has cached the old MAC address (ethernet hardware address) associated with the host's IP address. This cache will persist on the gateway network device until:
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: gettext

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 4]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+
+    <!-- Modal content-->
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Flush ARP system table</h4>
+      </div>
+      <div class="modal-body">
+        <p>
+        If you've changed the IP of a host - or setup a new host with the IP of an old one - you've probably got that the host has no network connectivity for a period of time. The router has cached the old MAC address (ethernet hardware address) associated with the host's IP address. This cache will persist on the gateway network device until:
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: gettext: {{ lang._('text') }}
indent

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 5]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+        </p>
+        <ul>
+            <li>
+        The ARP cache on the gateway network device expires.
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: same here

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 6]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+        </li>
+        <li>
+        You manually clear the ARP cache on the gateway network device.
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: same here

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 7]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+        </li>
+        </ul>
+        
+      </div>
+      <div class="modal-footer">
+        <button type="button" class="btn btn-default btn-close" data-dismiss="modal">Close</button>
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: only gettext here

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 8]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+        <button type="button" class="btn btn-default" data-toggle="modal" id="flush">Flush ARP tables</button>
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: ... and here

Comment (fabianfrz) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: + may look better if the T would be upcased

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 9]
```diff
@@ -61,13 +61,58 @@ POSSIBILITY OF SUCH DAMAGE.
...
+
+<!-- Modal -->
+<div id="flushModal" class="modal fade" role="dialog">
+  <div class="modal-dialog">
+
+    <!-- Modal content-->
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Flush ARP system table</h4>
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: gettext - see next comment

**Comment by peppelinux** [comment]
@fabianfrz Thank you for revision, I'll do better next time. Changes commited :+1:

**Comment by fabianfrz** [comment]
@peppelinux one is still left.

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt (hunk 9 above)
Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: fixed :+1:

**Comment by peppelinux** [comment]
You'll find it done in the last commit! :+1:

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt [hunk 10]
```diff
@@ -86,7 +86,7 @@ POSSIBILITY OF SUCH DAMAGE.
     <div class="modal-content">
       <div class="modal-header">
         <button type="button" class="close" data-dismiss="modal">&times;</button>
-        <h4 class="modal-title">Flush ARP system table</h4>
+    <h4 class="modal-title">{{ lang._('Flush ARP Table') }}</h4>
```
[the remark this answers is not in the document]
Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: fixed, my fault

File: src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php [hunk 11]
```diff
@@ -76,6 +76,22 @@ public function getArpAction()
...
+        if ($this->request->isPost()) {
+            $backend = new Backend();
+            $response = $backend->configdpRun("interface flush arp");
+            return $response;
+            }
+        else {
+            return array("message" => "error");
+            }
```
Comment (AdSchellevis) [inline src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/InterfaceController.php]: spacing seems off.

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt (hunk 2 above)
Comment (AdSchellevis) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: can you please simplify this code? a simple bootstrap dialog containing the ajaxCall(....) should be more then enough.

**Comment by peppelinux** [comment]
I'm very sorry about all these indentation errors, I code in this project during the night with tired eyes, you know. Thank you for your corrections, it's very important for me.

File: src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt (hunk 2 above)
Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: Ok, If it's important I will soon as possible

Comment (peppelinux) [reply src/opnsense/mvc/app/views/OPNsense/Diagnostics/arp.volt]: Done, you was right, it's much better then before

**Pull request merged**